### ISSUE TITLE
LongStorage seg fault

### ISSUE THREAD

**Issue by jonathantompson** [torch/torch7#166]
We probably shouldn't seg fault if a long storage is indexed...

>> x = torch.LongStorage(5,2)
>> print(x)

**Comment by soumith** [comment]
i actually dont know why it segfaults, what have you found when hunting the code? LongStorage(5,2) is not a valid constructor right...

**Comment by deltheil** [comment]
The seg fault occurs [here](https://github.com/torch/torch7/blob/7fd571f/lib/TH/generic/THStorage.c#L168).

Since the bounds are OK this is because the `data` pointer is invalid. Indeed looking at [the constructor](https://github.com/torch/torch7/blob/7fd571f/generic/Storage.c#L32-L38) we can see that `torch.LongStorage(5, 2)` has for effect to create a storage with a data pointer set to the address of the 2nd arg:

```C
/* (x) the 2nd arg is not the address of a pre-allocated memory zone! */
real *ptr = (real *)luaL_optlong(L, 2, 0); /* (x) */
storage = THStorage_(newWithData)(ptr, size);
```

Indeed this has been introduced by commit e5342e8:

> Experimental feature: creating Storages from raw pointers, in Lua.

So we can say that `torch.LongStorage(5,2)` is a misuse here.

**Comment by jonathantompson** [comment]
Right, it's definitely a misuse.  It accidentally came about because I
absentmindedly typed LongStorage rather than LongTensor... I was surprised
that we got a hard segfault, and this is something we probably should fix
(assert somewhere).

**Comment by soumith** [comment]
wow, cedric thanks for the hunting, forgot that it even existed.

First step is to document it.

@jonathantompson how do we suppose we do that? We have to trust the user creating the Storage in this case, as we cant verify if a pointer is valid. There is no check we can do without the check triggering a segfault afaik

**Comment by deltheil** [comment]
If the second parameter was explicitly a `cdata` object (instead of a number that holds the pointer address) don't you think it would be less error-prone? It would not avoid misuse (passing an irrelevant `cdata`, like wrong type or size) but it would be a bit more explicit.

What kind of uses do we have apart from doing something like:

```Lua
require 'torch'

local t = torch.FloatTensor(3):fill(0)
local s = torch.FloatStorage(3, tonumber(torch.data(t, true))) -- true=asnumber
```

Or:

```Lua
require 'torch'
local ffi = require 'ffi'

local d = ffi.new('float[?]', 3); ffi.fill(d, 3*ffi.sizeof('float'), 0)
local s = torch.FloatStorage(3, tonumber(ffi.cast('intptr_t', d)))
```

(both manipulate intermediate `cdata`-s)

**Comment by deltheil** [comment]
> First step is to document it.

Looking at it twice I can see it exists a piece of doc covered by [[result] data(tensor, [asnumber])](https://github.com/torch/torch7/blob/7fd571f/doc/tensor.md#result-datatensor-asnumber):

> Last, the pointer can be returned as a plain Lua number. This can be useful to share pointers between threads [...]

There is a corresponding example but IMHO there is a problem with it:

```Lua
> t = torch.randn(10)
> p = torch.data(t,true)
> s = torch.Storage(10, p) --> p is not a number but a cdata that wraps a `intptr_t`
> tt = torch.Tensor(s)
-- tt and t are a view on the same data. 
```

This line should be: `s = torch.Storage(10, tonumber(p))` (it fixes the problem but I'm not that sure it is that safe/clean to manipulate pointer addresses like that).

**Comment by soumith** [comment]
I think the reasonable middle ground is to check that the second argument is an actual boolean and not mistakenly a number (or any non Boolean type). I'll prep a patch for this

**Comment by deltheil** [comment]
> check that the second argument is an actual boolean and not mistakenly a number

You mean the second argument of `torch.data(self, asnumber)`? But this will not change the problem of passing a wrong address when creating a storage like `torch.LongStorage(5,2)`, exact?

So adding a piece of doc on the [storage constructors section](https://github.com/torch/torch7/blob/master/doc/storage.md#constructors-and-access-methods) to advertise this constructor (as you suggested) would be better than nothing.

But also: what do you think about fixing the [2nd problem pointed out above](https://github.com/torch/torch7/issues/166#issuecomment-77530752)? We could patch the doc:

```diff
diff --git a/doc/tensor.md b/doc/tensor.md
index ebcbde0..cae934e 100644
--- a/doc/tensor.md
+++ b/doc/tensor.md
@@ -2070,7 +2070,7 @@ gets freed, then the second points to nothing):
 
 ```lua
 t = torch.randn(10)
-p = torch.data(t,true)
+p = tonumber(torch.data(t,true))
 s = torch.Storage(10, p)
 tt = torch.Tensor(s)
 -- tt and t are a view on the same data. 
```

Or directly perform this cast on `torch.data` and `torch.cdata`:

```diff
diff --git a/FFI.lua b/FFI.lua
index b35770c..e07cdfc 100644
--- a/FFI.lua
+++ b/FFI.lua
@@ -163,7 +163,7 @@ typedef struct THRealTensor
       if not self then return nil; end
       local data = self:data()
       if asnumber then
-         return ffi.cast('intptr_t', data)
+         return tonumber(ffi.cast('intptr_t', data))
       else
          return data
       end
@@ -175,7 +175,7 @@ typedef struct THRealTensor
       if not self then return nil; end
       local cdata = self:cdata()
       if asnumber then
-         return ffi.cast('intptr_t', cdata)
+         return tonumber(ffi.cast('intptr_t', cdata))
       else
          return cdata
       end
```

**Comment by andreaskoepf** [comment]
I just ran into this problem while trying to 'reinterpret' a ShortTensor as ByteTensor. After reading in storage.md "The optional second argument ptr is a number whose value is a pointer to a memory chunk of size size*sizeof(TYPE) (for example coming from the torch.data() method)."  I thought  

```
a = torch.ShortStorage(100)
b = torch.ByteStorage(200, torch.data(a))
```

might work. But:

```
torch.data(a) == torch.data(b) -- returned false
```

Adding tonumber flag and the evil tonumber hack it works:

```
b = torch.ByteStorage(200, tonumber(torch.data(a, true)))
torch.data(a) == torch.data(b) -- returns true
```

But using a lua number to pass a pointer feels really wrong. A safe way to pass a pointer would be a special lua userdata pointer object. To convert form ffi cdata pointer to pointer-userdata a `torch.wrappointer' function could be used. A userdata pointer object could be processed by classic C API cwrap.

An alternative solution would be to create a "raw view" storage object with a C function called via ffi.
